### PR TITLE
fix(mariadb): do not automatically parse JSON fields by checking meta

### PR DESCRIPTION
**Pull request by lesion**
This is related to an https://github.com/sequelize/sequelize/pull/14800

The type of field returned by mariadb does not depend on the version of mariadb, but on the version of mariadb WHEN the table was created!

https://mariadb.com/kb/en/result-set-packets/#field-types
For a [JSON](https://mariadb.com/kb/en/result-set-packets/json) column, the column type field will be MYSQL_TYPE_STRING, but the extended type will indicate 'json'.

the connector use this field to decide to parse or not the response:
https://github.com/mariadb-corporation/mariadb-connector-nodejs/blob/59da962ce4ae6338bb14ed8854a70526b429911f/lib/cmd/decoder/text-decoder.js#L98

during a select the connector put fields details inside a "meta" key that could be used to check if returned value from the connector is already parsed or not:
https://github.com/mariadb-corporation/mariadb-connector-nodejs/blob/f25081715c9e1f24a10017e70787108d0d2ced01/documentation/promise-api.md#array-result-sets

a gist to test things:
https://gist.github.com/lesion/af0c7f1370ab6eb05cafdabd5b39ac33

### CODE REVIEW

**Comment by ephys** [comment]
Is there a way to add a regression test for this? I'm thinking create a text column using `queryInterface.createTable` but have the associated model use `DataTypes.JSON`?

`query.js` is likely to be completely rewritten in the future and a regression test would ensure we don't lose this logic

**Comment by lesion** [comment]
> Is there a way to add a regression test for this? I'm thinking create a text column using `queryInterface.createTable` but have the associated model use `DataTypes.JSON`?
> `query.js` is likely to be completely rewritten in the future and a regression test would ensure we don't lose this logic

I think we can test this by manually creating a table with a raw query using the two type of field mariadb uses for JSON and check if the results of a `findAll` are the same, could be something like:

```javascript
  await sequelize.query("DROP TABLE IF EXISTS test");
  await sequelize.query("CREATE TABLE test (id INTEGER AUTO_INCREMENT PRIMARY KEY, value1 longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL, value2 longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL CHECK(json_valid(value2)))")
  await sequelize.query("INSERT INTO test (value1, value2) VALUES(?, ?)", { replacements: [JSON.stringify({ key: 'value' }), JSON.stringify({ key: 'value' })] } )
  const Test = sequelize.define('test', {
    value1: DataTypes.JSON,
    value2: DataTypes.JSON,
  }, {
    freezeTableName: true,
    timestamps: false
   })

  const ret = await Test.findAll({raw: true})

  console.error(typeof ret[0].value1, typeof ret[0].value2)
  console.error(ret[0].value1, ret[0].value2)

/** output without patch
string object
{"key":"value"} { key: 'value' }
**/

/** output with patch
object object
{ key: 'value' } { key: 'value' }
**/
```

**Comment by lesion** [comment]
> Is there a way to add a regression test for this? I'm thinking create a text column using `queryInterface.createTable` but have the associated model use `DataTypes.JSON`?
> 
> `query.js` is likely to be completely rewritten in the future and a regression test would ensure we don't lose this logic

could we merge this? I wrote a unit test too https://github.com/sequelize/sequelize/pull/15704/files#diff-6a7c61f5188b4fa30990dd0dac3a0990087521e6a9d4431e10beecce7967573bR129-R153

Sorry to bother you, I'm worried because mariadb 10.8 is in debian stable so just and `apt upgrade` and this bug comes out

**Comment by WikiRik** [comment]
Can you resolve the conflicts with `main`? We'll take a look soon so we can merge

File: packages/core/test/integration/json.test.js [hunk 1]
```diff
@@ -126,6 +126,31 @@ describe('model', () => {
...
+        }, {
+          freezeTableName: true,
+          timestamps: false,
+        });
+
+        await this.Posts.create({ metaOldJSONtype: 'some text', metaNewJSONtype: 'some text' });
+
+        const posts = await this.Posts.findAll({ raw: true });
+        expect(posts[0].metaOldJSONtype).to.equal(posts[0].metaNewJSONtype);
+
```
Comment (WikiRik) [inline packages/core/test/integration/json.test.js:150]:
```suggestion
```

File: packages/core/test/integration/json.test.js [hunk 2]
```diff
@@ -126,6 +126,31 @@ describe('model', () => {
       });
     }
 
+    if (current.dialect.name === 'mariadb') {
+      // MariaDB does not support native JSON type, it uses longtext instead
```
Comment (WikiRik) [inline packages/core/test/integration/json.test.js:130]: To reduce the nesting, can you move this if-statement inside the test (and return if dialect !== mariadb)?
And move the comments inside the test as well

**Review by WikiRik — COMMENTED** [review]
Minor suggestions to keep the test suite a little bit cleaner but the change and test itself looks good to me

File: packages/core/test/integration/json.test.js (hunk 2 above)
Comment (lesion) [reply packages/core/test/integration/json.test.js:130]: I can do this but note that I've used the same syntax of the [preceding test](https://github.com/sequelize/sequelize/blob/b849ab47f59d0e08b766a029fee19ccd4fb80ea7/packages/core/test/integration/json.test.js#L112-L127)

**Review by WikiRik — DISMISSED** [review]

**Comment by lesion** [comment]
nope, I need to take a look at the new test structure

**Review by WikiRik — APPROVED** [review]
Good points from the linter

**Comment by WikiRik** [comment]
@ephys I'll give you a few days to look at it and then I'll merge

**Comment by ephys** [comment]
I'm fine with this change. I don't think it's the role of `query.js` to parse things but considering it's already here it's fine and we'll redesign query.js in a future update

**Comment by lesion** [comment]
this bug has hit again because the connector changed the name of the `dataTypeFormat` to `_dataTypeFormat` so the check is not ok anymore:
https://github.com/mariadb-corporation/mariadb-connector-nodejs/commit/dbd547268a5b06bed5a8294409a57255d166b3c4

the good news is that there is a method to check if it's a JSON field called `isDataTypeFormatJson` that can be used instead. I'll try to create a MR

**Comment by WikiRik** [comment]
If you could create a PR that's great! I know we aren't on the latest version yet, there was a types issue. I haven't checked recently if a new version was released that fixed it (that would unblock https://github.com/sequelize/sequelize/pull/18209 )

**Comment by lesion** [comment]
I don't know why the test for this specific issue still pass, seems that `handleJsonSelectQuery` is never called by the test, I made a gist to reproduce the issue https://gist.github.com/lesion/0b3e001f75c1fa89593ff365c2e931f7?permalink_comment_id=6216115#gistcomment-6216115 that its pretty the same as the test and is failing as expected, could you give me some help?

**Comment by lesion** [comment]
> If you could create a PR that's great! I know we aren't on the latest version yet, there was a types issue. I haven't checked recently if a new version was released that fixed it (that would unblock #18209 )

sorry for the noise I did not check #18209 and it will fix this issue as well -> https://github.com/sequelize/sequelize/pull/18209/changes#diff-2755d6fc997bec77c0ea727ff1fc508dce1addf1e634489b1fc976af100bfb14R204

the connector fixes the types issue in the last 3.5.3 https://github.com/mariadb-corporation/mariadb-connector-nodejs/issues/346#issuecomment-4509029415 so I will try to upgrade the version on your PR

**Pull request merged**